crates/ruff_formatter/src/format_element/document.rs

```diff
@@ -122,8 +122,12 @@ impl Document {
             expands
         }
 
-        let mut enclosing: Vec<Enclosing> = Vec::new();
-        let mut interned: FxHashMap<&Interned, bool> = FxHashMap::default();
+        let mut enclosing = Vec::with_capacity(if self.is_empty() {
+            0
+        } else {
+            self.len().ilog2() as usize
```

Comment:
This helped to improve performance quiet a bit. I haven't done any science but our groups very likely form a tree, the depth of a balanced binary tree is probably not the worst approximation.